porcupine-core/src/Control/Arrow/FoldA.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE Arrows #-}
+
+-- | This module defines the type 'FoldA', which is a generalization of 'Fold'
+-- from the `foldl` package.
+
+module Control.Arrow.FoldA
+  ( module Control.Foldl
+  , FoldA(..)
+  , Pair(..)
+  , generalizeA
+  , specializeA
+  , premapA
+  , postmapA
+  , prefilterA
+
+  , toTup
+  , fromTup
+  , (&&&&)
+  , (****)
+  ) where
+
+import Prelude hiding (id, (.))
+
+import Control.Arrow
+import Control.Category
+import Control.Foldl
+import Data.Profunctor
+
+
+ret :: (Arrow arr) => t -> arr a t
+ret = arr . const
+{-# INLINE ret #-}
+
+data Pair a b = Pair !a !b
+
+toTup :: Pair a b -> (a,b)
+toTup (Pair x y) = (x,y)
+{-# INLINE toTup #-}
+
+fromTup :: (a,b) -> Pair a b
+fromTup (x,y) = Pair x y
+{-# INLINE fromTup #-}
+
+applyP :: Pair (a -> b) a -> b
+applyP (Pair f x) = f x
+{-# INLINE applyP #-}
+
+uncurryP :: (a -> b -> c) -> Pair a b -> c
+uncurryP f (Pair x y) = f x y
+{-# INLINE uncurryP #-}
+
+(****) :: (Arrow a) => a b c -> a b' c' -> a (Pair b b') (Pair c c')
+a1 **** a2 =
+  arr toTup >>> (a1 *** a2) >>> arr fromTup
+{-# INLINE (****) #-}
+
+(&&&&) :: (Arrow a) => a b c -> a b c' -> a b (Pair c c')
+a1 &&&& a2 = a1 &&& a2 >>> arr fromTup
+{-# INLINE (&&&&) #-}
+
+secondP :: (Arrow a) => a c c' -> a (Pair b c) (Pair b c')
+secondP ar =
+  arr toTup >>> second ar >>> arr fromTup
+
+-- | If arr is (->), then 'FoldA' is just 'Fold'
+data FoldA arr a b =
+  forall x. FoldA (arr (Pair x a) x) (arr () x) (arr x b)
```

Comment:
My intuition is that the seed should also take an argument. It will probably prove useful. You'd have

```haskell
data FoldA arr g a b =
  forall x. FoldA (arr (Pair x a) x) (arr g x) (arr x b)
```

(g is for Γ, because we are reading ``g `arr` x`` as an `x` in the context `g`).

Reply:
I wondered about that. But what do you think could be the use for that?

Reply:
I'm not sure, yet. It's really not important as it's easily changed after the fact. But it's often useful in category theory, because many categories don't have a well-behaved unit. So it sort of stood out to me.